### PR TITLE
Add support for nullable min/max

### PR DESCRIPTION
**Pull request by faizandfurious**
Adds functions for min/max with pointers, allowing for nullable values.

### CODE REVIEW

File: mathutil.go [hunk 1]
```diff
@@ -638,6 +645,36 @@ func Min(a, b int) int {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := Max(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free
```go
if *a > *b {
        return a
}

return b
```

File: mathutil.go [hunk 2]
```diff
@@ -638,6 +645,36 @@ func Min(a, b int) int {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := Min(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 3]
```diff
@@ -683,6 +720,36 @@ func UMin(a, b uint) uint {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := UMax(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 4]
```diff
@@ -683,6 +720,36 @@ func UMin(a, b uint) uint {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := UMin(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 5]
```diff
@@ -728,6 +795,36 @@ func MinByte(a, b byte) byte {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxByte(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 6]
```diff
@@ -728,6 +795,36 @@ func MinByte(a, b byte) byte {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinByte(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 7]
```diff
@@ -773,6 +870,36 @@ func MinInt8(a, b int8) int8 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxInt8(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 8]
```diff
@@ -818,6 +945,36 @@ func MinUint16(a, b uint16) uint16 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxUint16(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 9]
```diff
@@ -818,6 +945,36 @@ func MinUint16(a, b uint16) uint16 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinUint16(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 10]
```diff
@@ -863,6 +1020,36 @@ func MinInt16(a, b int16) int16 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxInt16(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 11]
```diff
@@ -863,6 +1020,36 @@ func MinInt16(a, b int16) int16 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinInt16(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 12]
```diff
@@ -908,6 +1095,36 @@ func MinUint32(a, b uint32) uint32 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxUint32(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 13]
```diff
@@ -908,6 +1095,36 @@ func MinUint32(a, b uint32) uint32 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinUint32(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 14]
```diff
@@ -953,6 +1170,36 @@ func MinInt32(a, b int32) int32 {
...
+	if a == nil && b == nil {
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxInt32(*a, *b)
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 15]
```diff
@@ -953,6 +1170,36 @@ func MinInt32(a, b int32) int32 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinInt32(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 16]
```diff
@@ -998,6 +1245,36 @@ func MinUint64(a, b uint64) uint64 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxUint64(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 17]
```diff
@@ -998,6 +1245,36 @@ func MinUint64(a, b uint64) uint64 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinUint64(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 18]
```diff
@@ -1043,6 +1320,36 @@ func MinInt64(a, b int64) int64 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MaxInt64(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 19]
```diff
@@ -1043,6 +1320,36 @@ func MinInt64(a, b int64) int64 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinInt64(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

File: mathutil.go [hunk 20]
```diff
@@ -773,6 +870,36 @@ func MinInt8(a, b int8) int8 {
...
+		return nil
+	}
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+	r := MinInt8(*a, *b)
+	return &r
```
Comment (cznic) [inline mathutil.go]: Can be made allocation free

**Review by cznic — CHANGES_REQUESTED** [review]
Thanks for the PR. Please address the inline comments, thank you.

**Comment by faizandfurious** [comment]
Changes have been pushed.

**Comment by bojdell** [comment]
Exciting - thanks for contributing @faizandfurious! This looks good to me, I'll let @cznic have the final approval though

**Review by cznic — APPROVED** [review]
LGTM. Please commit to the master branch, thank you.

**Comment by cznic** [comment]
I believe it was grantad to you before the LGTM. Let me know if not, I'll send the collaborator's invite again (tomorrow). Maybe it fell into the spam folder?

**Comment by faizandfurious** [comment]
👍

**Pull request merged**